### ISSUE TITLE
Compiler doesn't generate flow union type when selecting interface fields

### ISSUE THREAD

**Issue by nirrek** [facebook/relay#2145]
Assume we have the following types:

```graphql
interface I { id: String! }
type A implements I { id: String! uniqueToA: String! }
type B implements I { id: String! uniqueToB: String! }
```

If we try to select the `id` field on the interface like so:
```graphql
graphql`
  fragment fragmentWithInterfaceField on I {
    __typename
    id
    ... on A {
      uniqueToA
    }
    ... on B {
      uniqueToB
    }
  }
`;
```

Then the compiler will generate the following flow type

```jsx
export type fragmentWithInterfaceField = {|
  +__typename: string;
  +id: string;
  +uniqueToA?: string;
  +uniqueToB?: string;
|};
```

This is not really what we want, we really want a union type here. There is, however, a workaround to get the union type we really wanted. If we were instead to add the `id` field to the fragment spread for every concrete type, then we would get a union type generated:
```graphql
graphql`
  fragment fragmentWithNoInterfaceField on I {
    __typename
    ... on A {
      id
      uniqueToA
    }
    ... on B {
      id
      uniqueToB
    }
  }
`;
```

```jsx
export type fragmentWithNoInterfaceField = {|
  +__typename: "A";
  +id: string;
  +uniqueToA: string;
|} | {|
  +__typename: "B";
  +id: string;
  +uniqueToB: string;
|} | {|
  // This will never be '%other', but we need some
  // value in case none of the concrete values match.
  +__typename: "%other";
|};
```

Where this workaround breaks down is when your schema has a lot of concrete types, but you aren't selecting any unique fields from most of them in the fragment. In these cases you have to write fragment spreads on all possible concrete types just to select the field that is on the interface.

```graphql
graphql`
  fragment fragmentWithNoInterfaceField on I {
    __typename
    ... on A {
      id
      uniqueToA
    }
    ... on B {
      id
      uniqueToB
    }
    # Having an interface should prevent us from having to do this :(
    ... on C {
      id
    }
    ... on D {
      id
    }
    ... on E {
      id
    }
    ... on F {
      id
    }
    ... on G {
      id
    }
    # etc.
  }
`;
```

So in summary, it would be preferable if selecting interface fields did not prevent the union type from being generated.

Here's a repo that demonstrates the above.
https://github.com/nirrek/relay-modern-union-type-interface-repro

**Comment by nirrek** [comment]
Actually nah, thinking about it a bit more, that union type wouldn't make much sense here. As we'd want to be able to access the interface field without necessarily having to switch on the `__typename`.

Closing for now until I have a better idea of what type would make more sense to be exported here.
